I14 fast slam

LordBismaya opened the pull request:
functionalized the fast slam function. put it in the estimation folder. The runfastSlam function calls the function.

LordBismaya committed:
Functionalised the FastSLAM code. Still have to move them to the respective folders. Code works with different motion and measurement models.

LordBismaya committed:
Modified the Fast Slam function.
- Modified Variable Names.
- Functionalized/Commented the Code.
- Added the centroid of the particles in the Plot.
- Commented wherever felt necessary.

LordBismaya committed:
Minor Modifications to the functions.

LordBismaya committed:
Update Particle WeCovariance Weight

LordBismaya committed:
Corrected covariance weights

LordBismaya committed:
Added FastSLAM2.0

LordBismaya committed:
TONS OF CHANGES. Major changes include FAST SLAM2.0 and functionalization of the whole code and addition of comments.

LordBismaya committed:
Add true state

LordBismaya committed:
Add Legend

Chiasung commented on matlab_simulation/05-estimation/ekf_fs.m:
```diff
@@ -0,0 +1,28 @@
+function [meanFeatP,covaFeatP,Hmu]=ekf_fs(predFeatPrior,obsFeat,meanFeat,covaFeat,measUncertainty,checkNewFeat)
+  if (checkNewFeat == 1) %nonobserved features are all 1;old features are 0;
+                % If new feature, initialize the features.
+                meanFeatP(1,1) = predFeatPrior(1,1)+obsFeat(1,1)*cos(obsFeat(2,1)+predFeatPrior(3,1));
```
If this linearised motion model can be the argument of fastSLAM(), the fastSLAM() function is wonderful.

Chiasung commented on matlab_simulation/05-estimation/fastSLAM.m:
```diff
@@ -0,0 +1,104 @@
+function[newPose,map,y,muParticleNew,covParticleNew,newParticleSet,meas_ind,newfeature,centroid_particles,w_new]...
+        =fastSLAM(oldPose,u,MotionNoise,MeasurementNoise,map,particleSet,muFeatOld,covFeatOld,w_new,rmax,thmax,timeInterval,totalParticles,newfeature)
```
Add usage and description of each argument here. That will be helpful for newbies.

Chiasung commented on matlab_simulation/05-estimation/runFastSLAM.m:
```diff
@@ -0,0 +1,61 @@
+% Author: Bismaya Sahoo.
+%% Initializations
+clear all; close all; clc;
+% Primary Initializations           
```
This program is great and well organized.  Only minor thing is regarding the code style. [Use alignment wherever it enhances readability](https://sites.google.com/site/matlabstyleguidelines/layout-documentation/white-space#TOC-Use-alignment-wherever-it-enhances-readability)

Chiasung commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m:
```diff
@@ -0,0 +1,172 @@
+    function newPose=motionUpdate_fs(oldPose,controlInput,MotionNoise,dt)
+    [QE, Qe] = eig(MotionNoise);n = length(MotionNoise(:,1)); 
+     e = QE*sqrt(Qe)*randn(n,1);
+    % Update robot state
```
If the motion model can also be the argument of fastSLAM(), this program is lovely:)

Chiasung commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m:
```diff
@@ -0,0 +1,172 @@
+               covFeat(:,:,i,particle),R,newfeature(i));
+           yw(2*(j-1)+1:2*j) = y(:,j);
+           hmuw(2*(j-1)+1:2*j) = Hmu;
+           Qw(2*(j-1)+1:2*j,2*(j-1)+1:2*j) = 2*covFeatPred(:,:,i,particle);
```
Just one quick question. What's the purpose of "2" in this equation-- "2*covFeatPred(:,:,i,particle)".

LordBismaya replied:
Yes. I did not understand that either. Ideally, it should have been..
H_feat_covFeatPred(:,:,i,particle)_H_feat'+measurementNoise;
where H_feat is the measurement jacobian with respect to the predicted features at that time instant and CovFeatPred is the covariance of that particle at that time step of a particular feature. 

I have made the changes in the function and it works. I shall push the updated code once the review is complete. Nice that you noticed..;-)

Rujie92 commented on matlab_simulation/05-estimation/runFastSLAM.m:
```diff
@@ -0,0 +1,61 @@
+dt=0.1; timeInterval=dt;%time interval
+finalTime=20;%final Time in sec
+MotionNoise=[0.001 0 0;0 0.001 0;0 0 0.0001];R=MotionNoise;
+MeasurementNoise=[0.02 0;0 0.02];Q=MeasurementNoise;
```
Great jobs, but I would suggest a single-column and aligned coding style with spaces besides signs like

MotionNoise = [ 1, 2, 3 ;
(space space..)                          4, 5, 6 ;
(space space..)                          7, 8, 9 ] ;
R = MotionNoise;

MesasurementNoise = [ 1, 2, 3 ;
(space space space.....)                                       4, 5, 6 ;
(space space space.....)                                       7, 8, 9 ] ;
R = MotionNoise;

Rujie92 commented on matlab_simulation/05-estimation/runFastSLAM.m:
```diff
@@ -0,0 +1,61 @@
+rmax = 10;
+thmax = pi/4;
+u=[1;0.3];
+%Fast Slam Specific Initializations
```
Separate lines with more "%" might be more visually clear and easy to understand.

`% Initiate the robot pose`
`pose(:,1) = poseInitial;`

`% xxxxx`
 `totalParticles=ParticleSize;`
 `particleSet=zeros(size(poseInitial,1),totalParticles);`

`% Mean of xxx`
 `muFeat=zeros(size(Q,1),noFeatures,totalParticles);`
 `muFeatPred=muFeat;`

`% Covariance of xxx`
 `covFeat=zeros(size(Q,1),size(Q,1),noFeatures,totalParticles);`
 `covFeatPred=covFeat;`

Rujie92 commented on matlab_simulation/05-estimation/runFastSLAM.m:
```diff
@@ -0,0 +1,61 @@
+        covFeatOld=covFeat;
+    end
+    %call the fast slam function
+    [newPose,map,y,muParticleNew,covParticleNew,newParticleSet,meas_ind,newfeature,centroid_particles,w_new]...
```
Suggest shortenning the naming of variables so we could see what are the imputs/outputs and how many are they.

By the way, maybe mu_Feat_old more ituitive to indicate the meaning of variables.

Rujie92 commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/ekf_fs.m:
```diff
@@ -0,0 +1,28 @@
+function [meanFeatP,covaFeatP,Hmu]=ekf_fs(predFeatPrior,obsFeat,meanFeat,covaFeat,measUncertainty,checkNewFeat)
```
Extra descriptions (What this function is used for, what are the inputs and outputs of the function and some other notices might be added) before coding in a function so that your codes could be easily viewed with little preliminary knowledge.

Rujie92 commented on matlab_simulation/05-estimation/runFastSLAM.m:
```diff
@@ -0,0 +1,61 @@
+    centroid(:,t)=centroid_particles';
+    
+    %Plot the Results
+    plot_fs(pose,map,y,muParticleNew,newParticleSet,t,meas_ind,totalParticles,noFeatures,newfeature,centroid);
```
Legends will be good explanationa for what is going on in the figure.

LordBismaya committed:
1.separate files for functions (unlike subfunctions used earlier),as per suggestion from reviewers.
2.Instructions on how to use.
3.added true robot pose, for reference
4.added error plots and rms error to the results
5.option to chose between random and static map

LordBismaya commented:
The code is updated after review.
1. Modified the Covariance weight update equation.(Resampling Step)
2. Added noise free state of the robot for reference in the plots.
3. legends in the plots.
4. Added Error plots and RMS errors
5. fastSLAM2.0 added.
6. Input and Output variables descriptions in each function.
7. Separate folder for fastSLAM in 05/Estimation directory
8. Individual functions (unlike subfunctions used earlier); as suggested by reviewers. This might help in future while making some of the functions common across the whole repo.
9. Centroid of the Particles in plots, showing mean of the estimated robot pose.
10. Instructions and ReadMe file included.
11. Comments wherever necessary. Variable names to enhance understanding.
12. Option to switch between random and static maps.(details in readme and instructions)

Chiasung commented on matlab_simulation/05-estimation/fastSLAM/func_fastSLAM2.m:
```diff
@@ -0,0 +1,80 @@
+           i = meas_ind(j);
+           
+%          %Sample Proposal (new in 2.0)
+           [mu_pose,cov_pose,H_pose,H_feat,pred_obs]=sample_proposal(...
```
Good work. func_fastSLAM & func_fastSLAM2 only slightly different. on this function call and Qw update. It's better to merge these two function into one and user can select the method by flag. That would be awesome.

LordBismaya commented:
Yes. That option had occurred to me. But I instead decided on keeping two separate m files for the two functions. I feel I have adequately mentioned the differences in comments to enable the future learner to clearly distinguish the differences between the two fairly easily.

Chiasung commented on matlab_simulation/05-estimation/fastSLAM/README.txt:
```diff
@@ -0,0 +1,11 @@
+
+5. All the functions contain descriptions of the input and output parameters
+
+NOTE: This example is created as a part of Assignment:3 in ME-780 Autonomous Mobile Robotics(Winter'16) course taught by Prof.Steven Waslander at the University of Waterloo. Reuse and redistribution of this code must be done after prior approval from him and the author.
```
I believe this is not open source license. May I ask for the prior approval to reuse this code in the future?

LordBismaya replied:
Haha...I had earlier read a similar ReadMe file with such a disclaimer/Important Note. This is just an attempt to fall into the habit of including a ReadMe/ license,but do feel free to use.:P

chutsu commented on matlab_simulation/05-estimation/fastSLAM/ekf_correct.m:
```diff
@@ -0,0 +1,25 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+function [meanFeatP,covaFeatP]=ekf_correct(muFeat,covFeat,obsFeat,H_feat,pred_obs,measUncertainty)
+        I = obsFeat(:,1)-pred_obs;
```
this only grabs the 1st observed feature?

LordBismaya replied:
No. If you look at the way the ekf_correct function is called, it loops through all of the observed features and processes them one by one.

chutsu commented on matlab_simulation/05-estimation/fastSLAM/featureSearch_fs.m:
```diff
@@ -0,0 +1,36 @@
+                end
+        end
+            function yes = inview(f,x, rmax, thmax)
+            % Checks for features currently in robot's view
```
functions within a function is a bad idea... also which inview function is line 18 calling then?

LordBismaya replied:
As long as it is working fine, could you please tell me why it is a bad idea? In opinion, it is a means of grouping modules together that perform similar actions. Here, in this case, the idea of the feature search is to search for the observed features and the "invew" function scans and locates those features. IMHO, it accomplishes a sub part of the whole objective of the feature_search function. 

Infact, In one of my earlier implementations, I had implemented the whole code as one main function with subfunctions. I only separated them as a different files so that they could be shared in future. Having said that, leaving the inview function within the feature search was deliberate.  Also, I love compact codes, hence, I have an inclination towards functions within functions..but thats just me. 

Also, if u read the official documentation of [MATLAB],(http://www.mathworks.com/help/matlab/matlab_prog/local-functions.html), you can see that local functions have precedence over other files and functions. And since, there are no other "invew" functions defined in the scope of fast slam, it should have been straightforward to understand which function it was calling.

chutsu replied:
"As long as it is working fine" is not a good measure for readable code. The fact that there is a global function `inview()` and you defining a local function **later** after `inview()` is first called confuses the reader, if the reader is not well versed in matlab's function precedence he/she will not know which version of `inview` will be called. People are lazy, they don't want to read the doc to find that out, neither do I. My suggestion is to either use the _global_ `inview`  function or atleast give the local one a different name.

Further more nested functions are generally not considered good practice, since if you have the justification for defining a function, it should be available for anyone to call, else it should not be a function. It is indeed a matter of taste, but I can assure you it is seldom used in general.

LordBismaya replied:
Ok. My bad, I did not realise that there was a global inview function. I understand your concerns. I shall update it to work with the global inview function later, as I am not sure if we are allowed to make changes to our code now.  

I'd be happy to take any more comments that you might have, and incorporate them after the evaluations are done. Thanks!

chutsu commented on matlab_simulation/05-estimation/fastSLAM/func_fastSLAM.m:
```diff
@@ -0,0 +1,78 @@
+        [newParticleSet,muParticleNew,covParticleNew]=raoBlackwell_fs(totalParticles,w,particleSetP,muFeatPred,covFeatPred);
+       
+        w_new=w;        
+end
```
consider adding spaces between mathematical operators and new lines between lines to increase readability, it is currently very dense.

LordBismaya replied:
ok.will do.good idea.

chutsu commented on matlab_simulation/05-estimation/fastSLAM/func_fastSLAM.m:
```diff
@@ -0,0 +1,78 @@
+
+
+function[newPose,y,muParticleNew,covParticleNew,newParticleSet,meas_ind,newfeature,w_new]...
+        =fastSLAM(oldPose,u,MotionNoise,MeasurementNoise,map,particleSet,muFeatOld,covFeatOld,rmax,thmax,timeInterval,totalParticles,newfeature,w)
```
Could use a struct here instead of having 14 arguments.

LordBismaya replied:
Noted..I could do that. Infact, this was my original plan. If you checkout Tim Bailey's original code, structs have been used. (https://openslam.org/bailey-slam.html) . But since, we were all up against a deadline, I didnt have enough time to update the whole code to a structure based format. And given that we had to stick to our deadline of having everything done by Apr 1st, I will go ahead and leave it as it is, since the code is working, and the variable names are descriptive enough. I plan update it with a structure format after the evaluations are done . :-)

LordBismaya committed:
Structures added to the code as per reviewer's comments. Plot function improved.Rounding done to the covariance matrix to prevent cholcov in mvnpdf function which crashed when the covaraince was very very low. All lose ends tied.
